drivers/wifi/nrfwifi/src/shim.c

```diff
@@ -0,0 +1,973 @@
+#include <zephyr/sys/__assert.h>
+
+#include "rpu_hw_if.h"
+#include "shim.h"
+#include "work.h"
+#include "timer.h"
+#include "osal_ops.h"
+#include "qspi_if.h"
+
+LOG_MODULE_REGISTER(wifi_nrf, CONFIG_WIFI_NRF70_LOG_LEVEL);
+
+struct zep_shim_intr_priv *intr_priv;
+
+static void *zep_shim_mem_alloc(size_t size)
+{
+	size = (size + 4) & 0xfffffffc;
+	return k_malloc(size);
+}
+
+static void *zep_shim_mem_zalloc(size_t size)
+{
+	size = (size + 4) & 0xfffffffc;
+	return k_calloc(size, sizeof(char));
+}
+
+static void *zep_shim_mem_cpy(void *dest, const void *src, size_t count)
+{
+	return memcpy(dest, src, count);
+}
+
+static void *zep_shim_mem_set(void *start, int val, size_t size)
+{
+	return memset(start, val, size);
+}
+
+static int zep_shim_mem_cmp(const void *addr1,
+			    const void *addr2,
+			    size_t size)
+{
+	return memcmp(addr1, addr2, size);
+}
+
+static unsigned int zep_shim_qspi_read_reg32(void *priv, unsigned long addr)
+{
+	unsigned int val;
+	struct zep_shim_bus_qspi_priv *qspi_priv = priv;
+	struct qspi_dev *dev;
+
+	dev = qspi_priv->qspi_dev;
+
+	if (addr < 0x0C0000) {
+		dev->hl_read(addr, &val, 4);
+	} else {
+		dev->read(addr, &val, 4);
+	}
+
+	return val;
+}
+
+static void zep_shim_qspi_write_reg32(void *priv, unsigned long addr, unsigned int val)
+{
+	struct zep_shim_bus_qspi_priv *qspi_priv = priv;
+	struct qspi_dev *dev;
+
+	dev = qspi_priv->qspi_dev;
+
+	dev->write(addr, &val, 4);
+}
+
+static void zep_shim_qspi_cpy_from(void *priv, void *dest, unsigned long addr, size_t count)
+{
+	struct zep_shim_bus_qspi_priv *qspi_priv = priv;
+	struct qspi_dev *dev;
+
+	dev = qspi_priv->qspi_dev;
+
+	if (count % 4 != 0) {
+		count = (count + 4) & 0xfffffffc;
+	}
+
+	if (addr < 0x0C0000) {
+		dev->hl_read(addr, dest, count);
+	} else {
+		dev->read(addr, dest, count);
+	}
+}
+
+static void zep_shim_qspi_cpy_to(void *priv, unsigned long addr, const void *src, size_t count)
+{
+	struct zep_shim_bus_qspi_priv *qspi_priv = priv;
+	struct qspi_dev *dev;
+
+	dev = qspi_priv->qspi_dev;
+
+	if (count % 4 != 0) {
+		count = (count + 4) & 0xfffffffc;
+	}
+
+	dev->write(addr, src, count);
+}
+
+static void *zep_shim_spinlock_alloc(void)
+{
+	struct k_sem *lock = NULL;
+
+	lock = k_malloc(sizeof(*lock));
+
+	if (!lock) {
+		LOG_ERR("%s: Unable to allocate memory for spinlock", __func__);
+	}
+
+	return lock;
+}
+
+static void zep_shim_spinlock_free(void *lock)
+{
+	k_free(lock);
+}
+
+static void zep_shim_spinlock_init(void *lock)
+{
+	k_sem_init(lock, 1, 1);
+}
+
+static void zep_shim_spinlock_take(void *lock)
+{
+	k_sem_take(lock, K_FOREVER);
+}
+
+static void zep_shim_spinlock_rel(void *lock)
+{
+	k_sem_give(lock);
+}
+
+static void zep_shim_spinlock_irq_take(void *lock, unsigned long *flags)
+{
+	k_sem_take(lock, K_FOREVER);
+}
+
+static void zep_shim_spinlock_irq_rel(void *lock, unsigned long *flags)
+{
+	k_sem_give(lock);
+}
+
+static int zep_shim_pr_dbg(const char *fmt, va_list args)
+{
+	char buf[80];
```

Comment:
If this is only used for debugging prints, perhaps just have it static, saves a bit stack usage.
Or have a global debug buffer and just using it in various debug prints, saves even more memory.

Reply:
global makes it tricky to make it thread safe, will make it static.